File: src/normalizer.ts
import Decimal from "decimal.js";

const UNIT_TO_KG: Record<string, Decimal> = {
  g: new Decimal("0.001"),
  kg: new Decimal("1"),
};

const UNIT_TO_L: Record<string, Decimal> = {
  ml: new Decimal("0.001"),
  cl: new Decimal("0.01"),
  l: new Decimal("1"),
};

const UNIT_ALIASES = new Set(["unit", "piece", "stuk"]);

export function normalizePrice(price: Decimal, quantity: Decimal, unit: string, target: string): Decimal {
  const u = unit.toLowerCase();
  const t = target.toLowerCase();
  if (quantity.lte(0)) {
    throw new Error("Quantity must be positive");
  }
  if (t === "kg" && u in UNIT_TO_KG) {
    return price.div(quantity.mul(UNIT_TO_KG[u]!)).toDecimalPlaces(2);
  }
  if (t === "l" && u in UNIT_TO_L) {
    return price.div(quantity.mul(UNIT_TO_L[u]!)).toDecimalPlaces(2);
  }
  if (t === "unit" && UNIT_ALIASES.has(u)) {
    return price.div(quantity).toDecimalPlaces(2);
  }
  throw new Error(`Incompatible units: ${unit} -> ${target}`);
}

export function addVat(netPrice: Decimal, vatRate: Decimal): Decimal {
  return netPrice.mul(new Decimal(1).plus(vatRate)).toDecimalPlaces(2);
}

File: src/normalizer.py
from decimal import Decimal

UNIT_TO_KG = {
    "g": Decimal("0.001"),
    "kg": Decimal("1"),
}
UNIT_TO_L = {
    "ml": Decimal("0.001"),
    "cl": Decimal("0.01"),
    "l": Decimal("1"),
}

def normalize_price(price: Decimal, quantity: Decimal, unit: str, target: str) -> Decimal:
    unit = unit.lower()
    target = target.lower()
    if quantity <= 0:
        raise ValueError("Quantity must be positive")
    if target == "kg" and unit in UNIT_TO_KG:
        return (price / (quantity * UNIT_TO_KG[unit])).quantize(Decimal("0.01"))
    if target == "l" and unit in UNIT_TO_L:
        return (price / (quantity * UNIT_TO_L[unit])).quantize(Decimal("0.01"))
    if target == "unit" and unit in {"unit", "piece", "stuk"}:
        return (price / quantity).quantize(Decimal("0.01"))
    raise ValueError(f"Incompatible units: {unit} -> {target}")

def add_vat(net_price: Decimal, vat_rate: Decimal) -> Decimal:
    return (net_price * (Decimal("1") + vat_rate)).quantize(Decimal("0.01"))
